crates/compiler/mono/src/inc_dec.rs

```diff
@@ -1,1554 +1,1193 @@
-            // we fall through when no join point is available
-            if let Some(jvars) = jp_live_vars.get(id) {
-                vars.extend(jvars);
-            }
-
-            vars
+            new_let!(newer_stmt)
         }
-
-        Switch {
-            cond_symbol,
-            branches,
-            default_branch,
-            ..
+        Expr::Array {
+            elem_layout: _,
+            elems,
         } => {
-            vars.insert(*cond_symbol);
-
-            for (_, _info, branch) in branches.iter() {
-                vars.extend(collect_stmt(branch, jp_live_vars, vars.clone()));
-            }
-
-            vars.extend(collect_stmt(default_branch.1, jp_live_vars, vars.clone()));
-
-            vars
+            // For an array creation, we insert all the used elements.
+            let new_let = new_let!(stmt);
+
+            inc_owned!(
+                elems.iter().filter_map(|element| match element {
+                    // Literal elements are not reference counted.
+                    ListLiteralElement::Literal(_) => None,
+                    // Symbol elements might be reference counted.
+                    ListLiteralElement::Symbol(symbol) => Some(*symbol),
+                }),
+                new_let
+            )
         }
-
-        Crash(m, _) => {
-            vars.insert(*m);
-            vars
+        Expr::Reuse { .. } | Expr::Reset { .. } | Expr::ResetRef { .. } => {
+            unreachable!("Reset(ref) and reuse should not exist at this point")
         }
     }
 }
 
-fn update_jp_live_vars(j: JoinPointId, ys: &[Param], v: &Stmt<'_>, m: &mut JPLiveVarMap) {
-    let j_live_vars = MutSet::default();
-    let mut j_live_vars = collect_stmt(v, m, j_live_vars);
+/**
+Insert increment statements for the given symbols compensating for the ownership.
+*/
+fn consume_and_insert_inc_stmts<'a>(
+    arena: &'a Bump,
+    environment: &mut RefcountEnvironment,
+    usage: impl IntoIterator<Item = (Symbol, u64)>,
+    continuation: &'a Stmt<'a>,
+) -> &'a Stmt<'a> {
+    usage
+        .into_iter()
+        .fold(continuation, |continuation, (symbol, usage_count)| {
+            consume_and_insert_inc_stmt(arena, environment, &symbol, usage_count, continuation)
+        })
+}
 
-    for param in ys {
-        j_live_vars.remove(&param.symbol);
-    }
+/**
+Insert an increment statement for the given symbol compensating for the ownership.
+*/
+fn consume_and_insert_inc_stmt<'a>(
+    arena: &'a Bump,
+    environment: &mut RefcountEnvironment,
+    symbol: &Symbol,
+    usage_count: u64,
+    continuation: &'a Stmt<'a>,
+) -> &'a Stmt<'a> {
+    let new_count = match environment.consume_rc_symbol(symbol) {
+        // If the symbol is borrowed, we need to increment the reference count for each usage.
+        Ownership::Borrowed => usage_count,
+        // If the symbol is owned, we need to increment the reference count for each usage except one.
+        Ownership::Owned => usage_count - 1,
```

Comment:
this cannot over/underflow?

Reply:
It should never be called with 0, added an assert.